AccessController - Add last access stack to errors

mbaker3 opened the pull request:
Add the stack of the last access operation to lifecycle error messages.

### What is the current behaviour?

When debugging a forgotten `Release/Async()` call or multiple `Acquire/Async()` calls in a stack it's difficult to track down the other method(s) that caused the contention or unexpected state.

### What is the new behaviour?

When `ENABLE_UNITY_COLLECTIONS_CHECKS` is present `AccessController` will record the stack at the end of each `Acquire/Async()` and `Release/Async()` call. When an `InvalidOperationException` is thrown it will include the stack of the previous call. This is usually the hint needed to figure out why the `AccessController` was in an unexpected state.

### What issues does this resolve?
 - None

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [ ] Yes <!-- If so, what are the migration considerations? -->
 - [x] No

jkeon commented on Scripts/Runtime/Job/AccessControl/AccessController.cs:
```diff
@@ -56,6 +56,11 @@ private enum AcquisitionState
 
         private AcquisitionState m_State;
 
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
```suggestion
#if ANVIL_DEBUG_SAFETY_EXPENSIVE
```

jkeon commented on Scripts/Runtime/Job/AccessControl/AccessController.cs:
```diff
@@ -307,6 +320,24 @@ private void ValidateReleaseState(JobHandle releaseAccessDependency)
+        [Conditional("ENABLE_UNITY_COLLECTIONS_CHECKS")]
+        private void CaptureAccessOperationStack()
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
```suggestion
#if ANVIL_DEBUG_SAFETY_EXPENSIVE
```

jkeon commented on Scripts/Runtime/Job/AccessControl/AccessController.cs:
```diff
@@ -307,6 +320,24 @@ private void ValidateReleaseState(JobHandle releaseAccessDependency)
+
+        private string GetLastAccessOperationStack()
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
```suggestion
#if ANVIL_DEBUG_SAFETY_EXPENSIVE
```

jkeon reviewed (changes requested):
Looks good but I would change the flag from COLLECTIONS_CHECKS to our more expensive debug check ANVIL_DEBUG_SAFETY_EXPENSIVE

mbaker3 commented:
Converted all of `AccessController` over to use the anvil defines and updated messages.
Now you'll get your safety checks only if `ANVIL_DEBUG_SAFETY` is defined. If an error is encountered the log message will direct the developer to enable `ANVIL_DEBUG_SAFETY_EXPENSIVE` to get the stack trace info.
